modules/apps/portal-search-elasticsearch7/portal-search-elasticsearch7-impl/src/main/java/com/liferay/portal/search/elasticsearch7/internal/bootstrap/ElasticsearchBootstrap.java

```diff
@@ -0,0 +1,328 @@
+
+		String javaExecutable = _getJavaExecutable(
+			new File(System.getProperty("java.home")));
+
+		List<String> arguments = new ArrayList<>();
+
+		// JVM arguments from config/jvm.options
+
+		_parseJVMArguments(
+			arguments, javaExecutable, new File(esPathConf, "jvm.options"),
+			esClasspath);
+
+		// Arguments in bin/elasticsearch script
+
+		arguments.add("-Des.path.home=".concat(_ES_HOME.getAbsolutePath()));
+		arguments.add("-Des.path.conf=".concat(esPathConf.getAbsolutePath()));
+
+		// TODO: Hard coded values from bin/elasticsearch
+
+		arguments.add("-Des.distribution.flavor=default");
+		arguments.add("-Des.distribution.type=tar");
+		arguments.add("-Des.bundled_jdk=true");
+
+		// Environments
+
+		Map<String, String> environments = new HashMap<>();
+
+		// ProcessExecutor cleans up default environment
+
+		environments.putAll(System.getenv());
+
+		// Elasticsearch requires HOSTNAME to be set
+
+		environments.put("HOSTNAME", hostname);
+
+		// Process executor classpath
+
+		String classpath = _createClasspath(esClasspath);
+
+		ProcessConfig.Builder processConfigBuilder =
+			new ProcessConfig.Builder();
+
+		return processConfigBuilder.setArguments(
+			arguments
+		).setBootstrapClassPath(
+			classpath
+		).setEnvironment(
+			environments
+		).setJavaExecutable(
+			javaExecutable
+		).setProcessLogConsumer(
+			processLog -> {
+				if (ProcessLog.Level.DEBUG == processLog.getLevel()) {
+					if (_log.isDebugEnabled()) {
+						_log.debug(
+							processLog.getMessage(), processLog.getThrowable());
+					}
+				}
+				else if (ProcessLog.Level.INFO == processLog.getLevel()) {
+					if (_log.isInfoEnabled()) {
+						_log.info(
+							processLog.getMessage(), processLog.getThrowable());
+					}
+				}
+				else if (ProcessLog.Level.WARN == processLog.getLevel()) {
+					if (_log.isWarnEnabled()) {
+						_log.warn(
+							processLog.getMessage(), processLog.getThrowable());
+					}
+				}
+				else {
+					_log.error(
+						processLog.getMessage(), processLog.getThrowable());
+				}
+			}
+		).setReactClassLoader(
+			ElasticsearchBootstrap.class.getClassLoader()
+		).setRuntimeClassPath(
+			classpath
+		).build();
+	}
+
+	private String _getJavaExecutable(File javaHome)
+		throws FileNotFoundException {
+
+		File bin = new File(javaHome, "bin");
+
+		File java = new File(bin, "java");
+
+		if (!java.exists() || !java.isFile()) {
+
+			// Windows
+
+			java = new File(bin, "java.exe");
+		}
+
+		if (java.exists() && java.isFile()) {
+			if (_log.isDebugEnabled()) {
+				_log.debug("Java executable: " + java.getAbsolutePath());
+			}
+
+			return java.getAbsolutePath();
+		}
+
+		throw new FileNotFoundException(
+			"Unable to find the java executable used to start Liferay");
+	}
+
+	private void _parseJVMArguments(
+			List<String> arguments, String javaExecutable, File jvmOptionsFile,
+			File esClasspath)
+		throws Exception {
+
+		// TODO: inline ES jvm.options logic instead of starting a process
```

Comment:
@dantewang I do not know why we have to start a new process here, it should be able to invoke the class by reflection, it should be seen by current bundle classloader cause we included all the source code.

Tina.

Reply:
@tinatian The jar which contains this class is not published as maven artifact, so we can only get it from the external ES classpath. This class has to be called via it's main method and it prints the parsing result directly into System.out.